pkg/agent/controller/traceflow/traceflow_controller.go

```diff
@@ -0,0 +1,332 @@
+	defer c.queue.Done(obj)
+
+	// We expect strings (Node name) to come off the workqueue.
+	if key, ok := obj.(string); !ok {
+		// As the item in the workqueue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen: enqueueTraceflow only enqueues strings.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncTraceflow(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		klog.Errorf("Error syncing Traceflow %s, Aborting. Error: %v", key, err)
+		c.queue.Forget(key)
+	}
+	time.Sleep(time.Duration(2) * time.Second)
+	return true
+}
+
+// syncTraceflow
+func (c *Controller) syncTraceflow(traceflowName string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing Traceflow for %s. (%v)", traceflowName, time.Since(startTime))
+	}()
+
+	tf, err := c.traceflowClient.OpsV1alpha1().Traceflows().Get(traceflowName, v1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	c.updateTagCache(tf)
+	return c.startTraceflow(tf)
+}
+
+// startTraceflow
+func (c *Controller) startTraceflow(tf *opsv1alpha1.Traceflow) error {
+	// Deploy flow entries for traceflow
+	err := c.deployFlowEntries(tf)
+	defer func() {
+		if err != nil {
+			c.errorTraceflowCRD(tf)
+		}
+	}()
+	if err != nil {
+		return err
+	}
+
+	// Inject packet if this Node is sender
+	time.Sleep(time.Duration(2) * time.Second)
+	return c.injectPacket(tf)
+}
+
+func (c *Controller) deployFlowEntries(tf *opsv1alpha1.Traceflow) error {
+	klog.Infof("Deploy flow entries for Traceflow %s", tf.Name)
+	return c.ofClient.InstallTraceflowFlows(tf.Status.DataplaneTag)
+}
+
+func (c *Controller) injectPacket(tf *opsv1alpha1.Traceflow) error {
+	namespaceName := tf.Spec.Source.Namespace
+	podName := tf.Spec.Source.Pod
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(podName, namespaceName)
+	// Pod not found in current Node, skip inject packet
+	if len(podInterfaces) == 0 {
+		return nil
+	}
+
+	// Update Traceflow phase to RUNNING
+	klog.Infof("Injecting packet for Traceflow %s", tf.Name)
+	c.senders[tf.Status.DataplaneTag] = *tf
+
+	// Calculate destination MAC/IP
+	dstMAC := ""
+	dstIP := tf.Spec.Destination.IP
+	if dstIP == "" {
+		dstPodInterfaces := c.interfaceStore.GetContainerInterfacesByPod(tf.Spec.Destination.Pod, tf.Spec.Destination.Namespace)
+		if len(dstPodInterfaces) > 0 {
+			dstMAC = dstPodInterfaces[0].MAC.String()
+			dstIP = dstPodInterfaces[0].IP.String()
+		} else {
+			dstPod, err := c.podLister.Pods(tf.Spec.Destination.Namespace).Get(tf.Spec.Destination.Pod)
+			if err != nil {
+				return err
+			}
+			// dstMAC is "" for globalVirtualMAC
+			dstIP = dstPod.Status.PodIP
+		}
+	}
+
+	if tf.Spec.Packet.IPHeader.Protocol == 0 {
```

Comment:
Please add a comments here to explain the default protocol.

Reply:
Added comment here.